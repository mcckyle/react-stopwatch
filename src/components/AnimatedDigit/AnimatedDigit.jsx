// src/components/AnimatedDigit.jsx

import React from 'react';
import { motion, AnimatePresence } from 'framer-motion';
import styles from "./AnimatedDigit.module.css";

function AnimatedDigit({ value }) {
  return (
    <div className={styles.digitContainer}>
      <AnimatePresence initial={false} mode="popLayout">
        <motion.div
          key={value}
          initial={{ y: 40, opacity: 0 }}
          animate={{ y: 0, opacity: 1 }}
          exit={{ y: -40, opacity: 0 }}
          transition={{ duration: 0.25, ease: "easeOut" }}
          className={styles.digit}
        >
          {value}
        </motion.div>
      </AnimatePresence>
    </div>
  );
}

export default AnimatedDigit;
