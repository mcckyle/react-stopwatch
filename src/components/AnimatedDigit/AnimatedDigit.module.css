/* Filename: AnimatedDigit.module.css */

/* ============================
   DIGIT CONTAINER STYLE.
=============================== */
.container
{
    display: inline-flex;
    justify-content: center;
    align-items: center;

    min-width: 1ch;
    overflow: hidden;
}

/* ============================
   DIGIT CHARACTER STYLE.
=============================== */
.digit
{
    font-family: "JetBrains Mono", monospace;
    font-weight: 700;
    font-size: clamp(2rem, 5vw, 3rem);
    color: var(--text);
    text-shadow: 0 0 6px rgba(212, 175, 55, 0.16);

    transition: text-shadow 0.2s ease;
}

/* DIGIT HOVER GLOW STYLE. */
.digit:hover
{
    text-shadow: 0 0 8px rgba(212, 175, 55, 0.3);
}

/* ==== LIGHT THEME OVERRIDES STYLES: ========================= */
html.light .digit
{
    color: #3a2c00;
    text-shadow: 0 0 6px rgba(234, 179, 8, 0.25);
}

html.light .digit:hover
{
    text-shadow: 0 0 9px rgba(234, 179, 8, 0.35);
}

/* ==== RESPONSIVE DESIGN STYLES: ========================= */
@media(max-width: 768px)
{
    .digit { font-size: clamp(1.6rem, 5vw, 2.4rem); }
}
