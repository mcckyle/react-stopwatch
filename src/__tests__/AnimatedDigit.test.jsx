//File name: AnimatedDigit.test.jsx
//Author: Kyle McColgan
//Date: 15 October 2025
//Description: This file contains the unit test suite for the AnimatedDigit component.

import React from "react";
import { render, screen } from "@testing-library/react";
import AnimatedDigit from "../components/AnimatedDigit/AnimatedDigit.jsx";

//Mock Framer Motion to simplify the animation behavior for testing purposes.
jest.mock("framer-motion", () => ({
    motion: {
        span: ({ children, ...props }) => <span {...props}>{children}</span>,
    },
    AnimatePresence: ({ children }) => <>{children}</>,
}));

describe("AnimatedDigit Component", () => {

    //Test #1: Basic render - ensures no crash.
    test("renders without crashing", () => {
        render(<AnimatedDigit value={0} />);
        expect(screen.getByText("0")).toBeInTheDocument();
    });

    //Test #2: Correctness - Correct digit content.
    test("displays the correct digit value", () => {
        render(<AnimatedDigit value={5} />);
        expect(screen.getByText("5")).toBeInTheDocument();
    });

    //Test #3: Correctness - Updates values when the prop changes.
    test("updates digit value when props change", () => {
        const { rerender } = render(<AnimatedDigit value={1} />);
        expect(screen.getByText("1")).toBeInTheDocument();


        rerender(<AnimatedDigit value={9} />);
        expect(screen.getByText("9")).toBeInTheDocument();
    });

    //Test #4: Structure - Includes a container wrapper. TO BE FIXED!!!
    test("renders a container div", () => {
        const { container } = render(<AnimatedDigit value={3} />);
        expect(container.firstChild.className).toContain("container");
    });

    //Test #5: Structure - Applies digit class for styling.
    test("applies the correct class to the animated digit", () => {
        const { container } = render(<AnimatedDigit value={4} />);
        const digit = container.querySelector("span");
        expect(digit.className).toContain("digit");
    });

    //Test #6: Accessibility - Uses aria-hidden for the animation element.
    test("marks the digit as aria-hidden", () => {
        const { container } = render(<AnimatedDigit value={7} />);
        const span = container.querySelector("span");
        expect(span).toHaveAttribute("aria-hidden", "true");
    });

    //Test #7: React Mechanics - Assigns unique key for each digit value.
    test("assigns a unique key based on the value", () => {
        const { container } = render(<AnimatedDigit value={8} />);
        const span = container.querySelector("span");
        expect(span.key).not.toBeNull;
    });

    //Test #8: React Mechanics - Supports multiple sequential renders.
    test("renders correctly across multiple rerenders", () => {
        const { rerender } = render(<AnimatedDigit value={2} />);
        rerender(<AnimatedDigit value={3} />);
        rerender(<AnimatedDigit value={4} />);
        expect(screen.getByText("4")).toBeInTheDocument();
    });

    //Test #9: React Mechanics - Does not render extra spans for the same value.
    test("renders only one span per value", () => {
        const { container } = render(<AnimatedDigit value={5} />);
        const spans = container.querySelectorAll("span");
        expect(spans.length).toBe(1);
    });

    //Test #10: Visual consistency - Snapshot test for consistent output.
    test("matches snapshot for digit 9", () => {
        const { asFragment } = render(<AnimatedDigit value={9} />);
        expect(asFragment()).toMatchSnapshot();
    });
});


