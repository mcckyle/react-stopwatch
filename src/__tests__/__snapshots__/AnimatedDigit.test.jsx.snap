// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AnimatedDigit Component matches snapshot for digit 9 1`] = `
<DocumentFragment>
  <div
    class="container"
  >
    <span
      animate="[object Object]"
      aria-hidden="true"
      class="digit"
      exit="[object Object]"
      initial="[object Object]"
      transition="[object Object]"
    >
      9
    </span>
  </div>
</DocumentFragment>
`;
